server/service/hosts.go

```diff
@@ -837,20 +840,63 @@ type hostsReportRequest struct {
 	Opts    fleet.HostListOptions `url:"host_options"`
 	LabelID *uint                 `query:"label_id,optional"`
 	Format  string                `query:"format"`
+	Columns string                `query:"columns,optional"`
 }
 
 type hostsReportResponse struct {
-	Hosts []*fleet.Host `json:"-"` // they get rendered explicitly, in csv
-	Err   error         `json:"error,omitempty"`
+	Columns []string        `json:"-"` // used to control the generated csv
+	Hosts   []*HostResponse `json:"-"` // they get rendered explicitly, in csv
+	Err     error           `json:"error,omitempty"`
 }
 
 func (r hostsReportResponse) error() error { return r.Err }
 
 func (r hostsReportResponse) hijackRender(ctx context.Context, w http.ResponseWriter) {
+	var buf bytes.Buffer
+	if err := gocsv.Marshal(r.Hosts, &buf); err != nil {
+		logging.WithErr(ctx, err)
+		http.Error(w, "Failed to generate CSV file", http.StatusInternalServerError)
+		return
+	}
+
+	// read back the CSV to filter out any unwanted columns
+	recs, err := csv.NewReader(&buf).ReadAll()
+	if err != nil {
+		logging.WithErr(ctx, err)
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprint(w, "Failed to generate CSV file")
+		return
+	}
+
+	var outRows [][]string
```

Comment:
Nit: How about a sanity check over `r.Columns`? Error out the request if an entry in `r.Columns` is unknown (not it `recs[0]`).

Or, just document in the API that unknown "columns" are ignored.